Update WiX toolset to 6.0.2

JoergAtGithub opened the pull request:
Closes: #15056 

Ensures for ARM64 from the begining that we generate valid installations. 
This hopefully will reduce the number of failed installation attemps, because the need for Admin rights is now correctly visualized and users see that they need elevated rights to install it.
Installer is generally smaller and faster.

JoergAtGithub committed:
Update the long time deprecated WiX toolset 3.14 to latest WiX toolset 6.0.1
Use the same WiX toolset installation on both Windows architectures
Fix the broken Multiuser Installation of the WiX 3.14 support in CPack

daschuer commented on .github/workflows/build.yml:
```diff
@@ -196,7 +196,7 @@ jobs:
           # This is a workaround for a SSL false positive in cmake 3.26.4
           # When downloading the manual. 3.21 is required for installing the
           # ANGLE Dlls via IMPORTED_RUNTIME_ARTIFACTS
```
Can you update/extend the comment? Even Ubuntu Jammy (min for Mixxx 2.5)  has cmake 3.22.

daschuer replied:
The comment is from here: 
https://github.com/mixxxdj/mixxx/pull/12197
Can you verify that downloading the manual works with CMake 3.30 

```suggestion
          # On Windows, we need at least CMake 3.30.x for WIX toolset >= 4.0 
```

daschuer commented on .github/workflows/build.yml:
```diff
@@ -374,13 +374,12 @@ jobs:
-          refreshenv
-          echo "WIX=$env:WIX" >> $env:GITHUB_ENV
+          dotnet.exe tool install --global wix --version 6.0.1
+          wix extension add --global WixToolset.UI.wixext/6.0.1
```
How has the situation with the WIX environment variable changed?

JoergAtGithub replied:
It is not needed anymore for WiX 4. As CI passes and generates proper MSI files, we can be sure, that the executable was found.

daschuer commented on CMakeLists.txt:
```diff
@@ -5056,6 +5056,8 @@ set(
 )
 
+# Use WiX toolset version >= 4.0 for building Windows installers.
+set(CPACK_WIX_VERSION 4)
```
We may also asume conributors have newer CMake versions. How is the situation regarding https://cmake.org/cmake/help/latest/policy/CMP0172.html#policy:CMP0172?

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.6' into Update_WiX_toolkit

JoergAtGithub committed:
Update/extend the comment about CMake minimum version:
-WiX >= v4 requires this version
-ANGLE is not suported by Qt6 anymore
-The download fail of the manual does not occur anymore

daschuer commented:
What happens if Mixxx 2.5 is already installed and you double click the new installer? Do we get a reasonable message?

daschuer commented:
I have not tested yet, but I think we need to detect the old upgrade GUID and prompt the user for manual uninstall. Than install with a new GUID. 

This seems to be a good template: 
https://github.com/Riskified/fluent-bit-riskified/blob/7b1c19a399905b411bf807c20a8754cff24ea1ff/cpack/wix/WIX.template.in#L3

Some more hints: 
https://stackoverflow.com/questions/17975898/prompt-to-uninstall-older-version-of-app-in-wix

JoergAtGithub committed:
Implemented a custom action FindAllRelatedMixxxInstallations as new DLL, that can be called inside of our MSI installer. FindAllRelatedMixxxInstallations detects old Mixxx installation with the illegal install scope "none"

JoergAtGithub commented:
I think I fixed this. The C# code section pointed by @daschuer does not generate MSI code, it is code that is called by MSI as Custom Action. I implemented this functionality as C++ DLL which is called as Custom Action in the MSI to get all previous Mixxx installations.

- The standad Windows API `FindRelatedProducts` returns only Installations where `UpgradeCode` and installation scope matches
- The custom `FindAllRelatedMixxxInstallations` returns all Installations where the `UpgradeCode` matches, regardless of the installation scope

This solves the issue with the old installations, where CPack set the invalid installation scope `none`.

daschuer commented on packaging/wix/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+# Finds old Mixxx installations with CPACK_WIX_INSTALL_SCOPE "none"
+# that the standard FindRelatedProducts action skips due to scope mismatch.
+if(WIN32)
+  add_library(MixxxUpgradeCA SHARED MixxxUpgradeCA.cpp)
```
I just noticed that we have normally all libs and cpps lower case. But I know that Windows treats that differently.
Was this a deliberate choice?

daschuer commented on CMakeLists.txt:
```diff
@@ -5080,6 +5080,8 @@ set(
 
+# Use WiX toolset version >= 4.0 for building Windows installers.
+set(CPACK_WIX_VERSION 4)
 set(CPACK_WIX_UPGRADE_GUID "921DC99C-4DCF-478D-B950-50685CB9E6BE")
```
What about this question? Do we have an advantage with a new ID?

daschuer reviewed (changes requested):
Great you have found a solution. I have added some nit picks for code style, basically. 
How it the DLL deployed? Does Wix care for it automatically? Maybe you can drop a comment about it.

JoergAtGithub committed:
Renamed cpp and dll to all lower case

JoergAtGithub committed:
Use const expression and explain their values in comments

JoergAtGithub committed:
Code-sign mixxxupgradecustomaction.dll

daschuer commented:
I have tested on a Windows 10 VM and got this error after installing on top of a 2.7 alpha install:  
<img width="409" height="178" alt="grafik" src="https://github.com/user-attachments/assets/2024161e-8bf7-495b-8a50-e68f0bf601e4" />

Really strange the start menu entry points to: 
C:\Users\dasch\Downloads\sign\mixxx.exe 
A folder I have created for testing purpose a time ago. 

The Mixxx.exe in the Program folder was missing, the other files are there. 

A "Repair" install repairs the installation. I have no clue what Microsoft is doing :-/

daschuer commented:
When downgrading to 2.5 from this 2.6 the user create a program folder of partially new and old files. 
This happens if a user downgrades form the new 2.6 to 2.5.

We need to advice the users that they uninstall 2.6 manually *Before* installing downgrading 2.5. With the old installer, the user was not able to downgrade and was prompted to uninstall the newer version manually.  
This is prompt is missing now.
 
Can we patch the installer for 2.5.5 based on Old Wix in the same way that a new 2.6 is found, that they can smoothy change between 2.5.5 and 2.6.0 at least?

For the already released 2.5.4 we can do nothing? 
Is this even worse the extra work? 

What do you think?

JoergAtGithub committed:
Use latest WiX release 6.0.2

daschuer reviewed (approved):
The code looks good. I read a bit and while the finding with the random mixxx.exe is a severe issue. It is unlikely caused by you changes. 

The question remains: do we want to add a similar patch to the 2.5 branch detecting the new installation?

daschuer reviewed (approved):
LGTM, thank you.
